### ISSUE TITLE
Fallback after data node downtime

### ISSUE THREAD

**Issue by painter1** [Prodiguer/synda#108]
Recently IPSL's data nodes were down.  My running replication job tried every IPSL file at gsiftp://gridftp.ipsl.upmc.fr.  When it failed, it retried at http://vesg.ipsl.upmc.fr.  When that failed, every file went to 'error' status.  That is exactly the way that Synda should work.

When the data nodes came back up, I issued a "synda retry".   The data came, but from http://vesg... at 30 MB/s rather than gsiftp://gridftp... at 100 MB/s.  I think that this happened because the downtime fallback left each file's url slot (in the database) with the http://vesg... value.  There was no way for Synda to go back to the preferred server.

If I'm right about where the problem is, we could handle this kind of situation by saving the original url, or by repeating the search after a retry.  Each solution is messy in its own way.

This is not an urgent issue because the problem does not prevent downloads; it merely affects performance.

**Comment by painter1** [comment]
I pushed some updates to my 'more_fallbacks' branch.  I don't know the status of the branch now.  But it has all the changes which I have been using to download hundreds of TB in the last couple of months, and I just tested it in a Synda version which is independent of the other local changes I have made.

There are some issues which I think need to be resolved before we merge this into the master branch and distribute it widely.

1. There are two changes needed in the database.  I do not presently have a plan for how to ensure that users' databases incorporate the changes:
(i) A new table, failed_url.  This simply keeps track of a list of urls for which downloads failed, in a way which makes the list available to all threads.  Something like this is necessary because there can be more complexity in the logic of fallbacks when you are doing something other than gsiftp->http->'give up'.  There may be a better solution, but I don't have one at the moment.
(ii) a new column, searchapi_host, in the file table.  Actually this is needed simply to fix a minor bug in the previous version of sdnexturl.py:  the user's choice of searchapi_host in the selection file should apply to sdnexturl, but doesn't, and can't unless the choice is remembered somehow.

2.  When many data nodes may be chosen, you have to choose one somehow.  I hard-wired a priority list into sdnexturl.py.  Other people may have different priorities!  We should have a more flexible way to choose data node priorities.  This is not a showstopper:  we can figure it out later.  Actually, data node priorities might be a useful feature in the _initial_ selection of a data node to download from.

3. Other issues which we should consider:  Is it ok to always prefer gsiftp over http?  I now think so, because it is ESGF policy that every data node should offer both. 

4. The other issues I brought up in my previous pull request don't matter now in my opinion:  I now longer am deleting the character set specification (although I still don't like it) and I think the logging is not excessive at present.

**Comment by AtefBN** [comment]
Thank you Jeff for creating and now updating this issue/thread.
No need to say I think that the feature is of course an appreciated enhancement to the existing Synda code base. But you're right we need to settle these points first.
Now I'm speaking for myself and not the entire IPSL, @SebastienDenvil might have other opinions on this and his input will also be appreciated. 

1. These changes in the db should be easy to implement in newer versions of Synda, either using a script appended to the new package or through sqlalchemy. However thinking about this now, I wonder if this will cause issues for users with older versions of synda. 
If we implement the db schema migration when updating SDT from 3.9 to 3.9.1 but say one user skips this iteration and goes straight to a future synda 4.0, there will be inconsistency in the database. I wonder how best to deal with this (please note this is me thinking out loud)

2. Priorities will always have to be hardcoded somewhere I think. We could think of implementing a selection "intelligence" within synda that takes the responsibility of choosing from the user either through learning the network or by simple punctual testing prior to downloading. But either way, I rather tackle this in the future. 

3. I personally have no say about this, I'm open to either really, could join 2. in having a smarter Synda doing the choices. 

To sum up, the pressing issue of ingesting this PR in Synda would be to figure out this mini DB update.
I'm confident there have been past db updates within synda and @SebastienDenvil can probably weigh in on this.

**Comment by SebastienDenvil** [comment]
@AtefBN yes there has been db update in the past and it was taken care of by update procedures. A bit tedious to manage honeslty. This was done prior to the rpm based synda.

This will be more tedious to manage if we have multiple schema updates in the future. We will need to be able to update from any version to the latest. When it starts synda could check the DB schema and act accordingly.

And by the way thanks @painter1 the fallback after a data node downtime is indeed important.

**Comment by painter1** [comment]
Looking at sddbversion.py, I think the solution is to change the Synda and database versions (what do you think of calling it 3.9.1?).  Then I would add some upgrade code in sddbversion.py.  I'll do this soon and make a separate pull request for it.

**Comment by AtefBN** [comment]
Yup thinking to go 3.9.1, changes don't warrant going to 4.0 that I want to keep for the python 3 release.

**Comment by painter1** [comment]
I ended out making it version 3.10 - not because of any particular preference, but because 3.9.1 didn't work at the time.  Then I changed sddbversion* to support either 3.10 or 3.9.1.  The changes to sdapp.py, sddbversion.py, and sddbversionutils.py are now in two branches: more_fallbacks and version310.  We already have a pull request for more_fallbacks (I think it's now ready to be approved, but form your own opinion).  In a moment I'll add a separate pull request for version310.

**Comment by painter1** [comment]
It looks like my more_fallbacks branch had already been merged in.  So you can ignore the change I just made to it, it's the same as my pull request on the version310 branch.
